File: src/main/kotlin/com/dokhabackend/dokha/config/security/JwtTokenUtil.java
package com.dokhabackend.dokha.config.security;

import com.dokhabackend.dokha.entity.User;
import io.jsonwebtoken.Claims;
import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.SignatureAlgorithm;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.stereotype.Component;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Date;
import java.util.function.Function;

/**
 * @author v.butuzov
 * Утилиты для jwt
 */
@Component
public class JwtTokenUtil implements Serializable {

    @Value("${jwt.accessTokenValidityMillisecond}")
    private long ACCESS_TOKEN_VALIDITY_SECONDS;

    @Value("${jwt.signingKey}")
    private String SIGNING_KEY;

    public String getUsernameFromToken(String token) {
        return getClaimFromToken(token, Claims::getSubject);
    }

    private Date getExpirationDateFromToken(String token) {
        return getClaimFromToken(token, Claims::getExpiration);
    }

    private <T> T getClaimFromToken(String token, Function<Claims, T> claimsResolver) {
        final Claims claims = getAllClaimsFromToken(token);
        return claimsResolver.apply(claims);
    }

    private Claims getAllClaimsFromToken(String token) {
        try {
            return Jwts.parser()
                    .setSigningKey(SIGNING_KEY)
                    .parseClaimsJws(token)
                    .getBody();
        } catch (Exception e) {
            throw new IllegalArgumentException("Не верный токен");
        }

    }

    private Boolean isTokenExpired(String token) {
        final Date expiration = getExpirationDateFromToken(token);
        return expiration.before(new Date());
    }

    public String generateToken(User user) {
        return doGenerateToken(user.getLogin(), user.getRoles().toString());
    }

    private String doGenerateToken(String subject, String role) {
        Collection<SimpleGrantedAuthority> simpleGrantedAuthorities = new ArrayList<>();
        simpleGrantedAuthorities.add(new SimpleGrantedAuthority(role));
        Claims claims = Jwts.claims().setSubject(subject);
        claims.put("role", simpleGrantedAuthorities);

        return Jwts.builder()
                .setClaims(claims)
                .setIssuer("dokha")
                .signWith(SignatureAlgorithm.HS256, SIGNING_KEY)
                .setIssuedAt(new Date(System.currentTimeMillis()))
                .setExpiration(new Date(System.currentTimeMillis() + ACCESS_TOKEN_VALIDITY_SECONDS * 1000))
                .compact();
    }

    public Boolean validateToken(String token, UserDetails userDetails) {
        return (getUsernameFromToken(token).equals(userDetails.getUsername()) && !isTokenExpired(token));
    }

}

File: src/main/kotlin/com/dokhabackend/dokha/config/security/JwtTokenUtil.kt
package com.dokhabackend.dokha.config.security

import com.dokhabackend.dokha.entity.User
import io.jsonwebtoken.Claims
import io.jsonwebtoken.Jwts
import io.jsonwebtoken.SignatureAlgorithm
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.beans.factory.annotation.Value
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.stereotype.Component
import java.util.*
import java.util.function.Function

/**
 * @author v.butuzov
 * Утилиты для jwt
 */
@Component
//Спринг пьян и не инжектить сюда черезе @ConfigProps/@Value
class JwtTokenUtil(val jwtProperties: JwtProperties) {

    fun getUsernameFromToken(token: String): String = getClaimFromToken(token, Function { it.subject })

    private fun getExpirationDateFromToken(token: String): Date {
        return getClaimFromToken(token, Function { it.expiration })
    }

    private fun <T> getClaimFromToken(token: String, claimsResolver: Function<Claims, T>): T {
        val claims = getAllClaimsFromToken(token)
        return claimsResolver.apply(claims)
    }

    private fun getAllClaimsFromToken(token: String): Claims {
        try {
            return Jwts.parser()
                    .setSigningKey(jwtProperties.signingKey)
                    .parseClaimsJws(token)
                    .body
        } catch (e: Exception) {
            throw IllegalArgumentException("Не верный токен")
        }

    }

    private fun isTokenExpired(token: String): Boolean {
        val expiration = getExpirationDateFromToken(token)
        return expiration.before(Date())
    }

    fun generateToken(user: User): String {
        return doGenerateToken(user.login, user.roles.toString())
    }

    private fun doGenerateToken(subject: String, role: String): String {
        val simpleGrantedAuthorities = ArrayList<SimpleGrantedAuthority>()
        simpleGrantedAuthorities.add(SimpleGrantedAuthority(role))
        val claims = Jwts.claims().setSubject(subject)
        claims["role"] = simpleGrantedAuthorities

        return Jwts.builder()
                .setClaims(claims)
                .setIssuer("dokha")
                .signWith(SignatureAlgorithm.HS256, jwtProperties.signingKey)
                .setIssuedAt(Date(System.currentTimeMillis()))
                .setExpiration(Date(System.currentTimeMillis() + jwtProperties.accessTokenValidityMillisecond * 1000))
                .compact()
    }

    fun validateToken(token: String, userDetails: UserDetails): Boolean? {
        return getUsernameFromToken(token) == userDetails.username && !isTokenExpired(token)
    }
}